test/libsolidity/GasTest.cpp

```diff
@@ -0,0 +1,222 @@
+		++_it;
+	return u256(std::string(begin, _it));
+}
+
+}
+
+GasTest::GasTest(string const& _filename)
+{
+	ifstream file(_filename);
+	if (!file)
+		BOOST_THROW_EXCEPTION(runtime_error("Cannot open test contract: \"" + _filename + "\"."));
+	file.exceptions(ios::badbit);
+
+	m_source = parseSourceAndSettings(file);
+
+	if (m_settings.count("optimize"))
+	{
+		m_optimise = true;
+		m_validatedSettings["optimize"] = "true";
+		m_settings.erase("optimize");
+	}
+	if (m_settings.count("optimize-yul"))
+	{
+		m_optimiseYul = true;
+		m_validatedSettings["optimize-yul"] = "true";
+		m_settings.erase("optimize-yul");
+	}
+	if (m_settings.count("optimize-runs"))
+	{
+		m_optimiseRuns = stoul(m_settings["optimize-runs"]);
+		m_validatedSettings["optimize-runs"] = m_settings["optimize-runs"];
+		m_settings.erase("optimize-runs");
+	}
+
+	parseExpectations(file);
+}
+
+void GasTest::parseExpectations(std::istream& _stream)
+{
+	std::map<std::string, std::string>* currentKind = nullptr;
+	std::string line;
+
+	while (getline(_stream, line))
+		if (boost::starts_with(line, "// creation:"))
+		{
+			auto it = line.begin() + 12;
+			skipWhitespace(it, line.end());
+			m_creationCost.executionCost = parseGasCost(it, line.end());
+			skipWhitespace(it, line.end());
+			if (*it++ != '+')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.codeDepositCost = parseGasCost(it, line.end());
+			skipWhitespace(it, line.end());
+			if (*it++ != '=')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.totalCost = parseGasCost(it, line.end());
+		}
+		else if (line == "// external:")
+			currentKind = &m_externalFunctionCosts;
+		else if (line == "// internal:")
+			currentKind = &m_internalFunctionCosts;
+		else if (!currentKind)
+			BOOST_THROW_EXCEPTION(runtime_error("No function kind specified. Expected \"external:\" or \"internal:\"."));
+		else
+		{
+			if (!boost::starts_with(line, "// "))
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"// \"."));
+			auto it = line.begin() + 3;
+			skipWhitespace(it, line.end());
+			auto functionNameBegin = it;
+			while (it != line.end() && *it != ':')
+				++it;
+			std::string functionName(functionNameBegin, it);
+			if (functionName == "fallback")
+				functionName.clear();
+			expect(it, line.end(), ':');
+			skipWhitespace(it, line.end());
+			if (it == line.end())
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected gas cost."));
+			(*currentKind)[functionName] = std::string(it, line.end());
+		}
+}
+
+void GasTest::printUpdatedExpectations(std::ostream& _stream, std::string const& _linePrefix) const
+{
+	Json::Value estimates = compiler().gasEstimates(compiler().lastContractName());
+	_stream << _linePrefix << "creation: "
+			<< estimates["creation"]["executionCost"].asString() << " + "
```

Comment:
too much indentation - and should also break after each `<<`... (or you introduce `(...)`)